openvdb/openvdb/tools/Merge.h

```diff
@@ -501,11 +501,52 @@ struct UnallocatedBuffer<BufferT, bool>
 template <typename TreeT, bool Union>
 bool CsgUnionOrIntersectionOp<TreeT, Union>::operator()(RootT& root, size_t) const
 {
+    const bool Intersect = !Union;
+
     if (this->empty())     return false;
 
     // store the background value
     if (!mBackground)   mBackground = &root.background();
 
+    // does the key exist in the root node?
+    auto keyExistsInRoot = [&](const Coord& key) -> bool
+    {
+        return root.getValueDepth(key) > -1;
+    };
+
+    // does the key exist in all merge tree root nodes?
+    auto keyExistsInAllTrees = [&](const Coord& key) -> bool
+    {
+        for (TreeToMerge<TreeT>& mergeTree : mTreesToMerge) {
+            const auto* mergeRoot = mergeTree.rootPtr();
+            if (!mergeRoot)                             return false;
+            if (mergeRoot->getValueDepth(key) == -1)    return false;
+        }
+        return true;
+    };
+
+    // delete any background tiles
+    root.eraseBackgroundTiles();
+
+    // for intersection, delete any root node keys that are not present in all trees
+    if (Intersect) {
+        // find all tile coordinates to delete
+        std::vector<Coord> toDelete;
+        for (auto valueIter = root.cbeginValueAll(); valueIter; ++valueIter) {
+            const Coord& key = valueIter.getCoord();
+            if (!keyExistsInAllTrees(key))   toDelete.push_back(key);
+        }
+        // find all child coordinates to delete
+        for (auto childIter = root.cbeginChildOn(); childIter; ++childIter) {
+            const Coord& key = childIter.getCoord();
+            if (!keyExistsInAllTrees(key))   toDelete.push_back(key);
+        }
+        // only mechanism to delete elements in root node is to delete background tiles,
+        // so insert background tiles (which will replace any child nodes) and then delete
+        for (Coord& key : toDelete)     root.addTile(key, *mBackground, false);
+        root.eraseBackgroundTiles();
```

Comment:
This is also called at the end of this function - do you need to call it twice?

Reply:
Yes, later portions of this routine check if the key exists in the target tree and then skips any merges if not. However, where the key is valid, it is still possible to merge in a background tile (as this method isn't first called on the source trees), so it needs to be called at the end again to ensure that they all finally get removed.

Typically these root node hash tables are very small, so this method is not expensive.